internal/client/db/db.go

```diff
@@ -0,0 +1,585 @@
+}
+
+// Decrease reference count for pinned item and remove if reference count is 0.
+func (bdb *blockchainDB[H, N, E, Header]) unpin(hash H) {
+	bdb.pinnedBlocksCacheMtx.Lock()
+	defer bdb.pinnedBlocksCacheMtx.Unlock()
+	bdb.pinnedBlocksCache.Unpin(hash)
+}
+
+func (bdb *blockchainDB[H, N, E, Header]) justificationsUncached(hash H) (runtime.Justifications, error) {
+	blockID := generic.NewBlockID[H, N](generic.BlockIDHash[H]{Inner: hash})
+	justificationsBytes, err := readDB[H, N](bdb.db, columns.KeyLookup, columns.Justifications, blockID)
+	if err != nil {
+		return nil, err
+	}
+	if justificationsBytes != nil {
+		var justifications runtime.Justifications
+		err := scale.Unmarshal(justificationsBytes, &justifications)
+		if err != nil {
+			return nil, err
+		}
+		return justifications, nil
+	}
+	return nil, nil
+}
+
+func (bdb *blockchainDB[H, N, E, Header]) bodyUncached(hash H) ([]runtime.Extrinsic, error) {
+	blockID := generic.NewBlockID[H, N](generic.BlockIDHash[H]{Inner: hash})
+	bodyBytes, err := readDB[H, N](bdb.db, columns.KeyLookup, columns.Body, blockID)
+	if err != nil {
+		return nil, err
+	}
+	if bodyBytes != nil {
+		var extrinsics []E
+		err := scale.Unmarshal(bodyBytes, &extrinsics)
+		if err != nil {
+			return nil, err
+		}
+		var body []runtime.Extrinsic
+		for _, e := range extrinsics {
+			body = append(body, e)
+		}
+		return body, nil
+	}
+
+	indexBytes, err := readDB[H, N](bdb.db, columns.KeyLookup, columns.BodyIndex, blockID)
+	if err != nil {
+		return nil, err
+	}
+	if indexBytes == nil {
+		return nil, nil
+	}
+	var index []dbExtrinsic[E]
+	err = scale.Unmarshal(indexBytes, index)
+	if err != nil {
+		return nil, err
+	}
+	var body []runtime.Extrinsic
+	for _, ex := range index {
+		dbex, err := ex.Value()
+		if err != nil {
+			return nil, err
+		}
+		switch dbex := dbex.(type) {
+		case dbExtrinsicIndexed:
+			t := bdb.db.Get(columns.Transaction, hash.Bytes())
+			if t != nil {
+				input := newJoinInput(dbex.Header, t)
+				var ex E
+				err := scale.Unmarshal(input.Bytes(), &ex)
+				if err != nil {
+					return nil, fmt.Errorf("Error decoding indexed extrinsic: %w", err)
+				}
+				body = append(body, ex)
+			} else {
+				return nil, fmt.Errorf("Missing indexed transaction %v", hash)
+			}
+		case dbExtrinsicFull[E]:
+			body = append(body, dbex.Extrinsic)
+		}
+	}
+	return body, nil
+}
+
+func (bdb *blockchainDB[H, N, E, Header]) cacheHeader(hash H, header *runtime.Header[N, H]) {
+	bdb.headerCache.Put(hash, header)
+	for bdb.headerCache.Size() > numCachedHeaders {
+		iterator := bdb.headerCache.Iterator()
+		if !iterator.First() {
+			panic("huh?")
+		}
+		bdb.headerCache.Remove(iterator.Key())
+	}
+}
+
+func (bdb *blockchainDB[H, N, E, Header]) Header(hash H) (runtime.Header[N, H], error) {
+	bdb.headerCacheMtx.Lock()
+	defer bdb.headerCacheMtx.Unlock()
+	val, ok := bdb.headerCache.Get(hash)
+	if ok {
+		// TODO: create issue to fork linkedhashmap, and add cache.get_refresh(&hash)
```

Comment:
Has an issue been created for this?

Reply:
Also perhaps this is not worth maintaining a fork over. Will there be significant differences?

Reply:
```
bdb.headerCache.Remove(hash)
bdb.headerCache.Put(hash, val)
```
is essentially the same as `get_refresh`.

Reply:
Sounds good. It seems there are a few tasks that are being revisited after the completion of db.Backend, are these documented? Perhaps we can create an issue with a task list of things to return to once db.Backend is completed so nothing falls through the cracks

Reply:
I would say there's 2 more PRs which will complete the `db.Backend`, I'll probably address them in #3902.  I'll be sure to track any thing I need to revisit by linking these threads in #3902 as well as through TODOs in the code.